src/libs/actions/Session/index.ts

```diff
@@ -107,15 +183,59 @@ function signOut() {
  * Checks if the account is an anonymous account.
  */
 function isAnonymousUser(): boolean {
-    return sessionAuthTokenType === CONST.AUTH_TOKEN_TYPE.ANONYMOUS;
+    return session.authTokenType === CONST.AUTH_TOKEN_TYPES.ANONYMOUS;
+}
+
+function hasStashedSession(): boolean {
+    return Boolean(stashedSession.authToken && stashedCredentials.autoGeneratedLogin && stashedCredentials.autoGeneratedLogin !== '');
 }
 
-function signOutAndRedirectToSignIn(shouldReplaceCurrentScreen?: boolean) {
+function signOutAndRedirectToSignIn(shouldReplaceCurrentScreen?: boolean, stashSession?: boolean) {
     Log.info('Redirecting to Sign In because signOut() was called');
     hideContextMenu(false);
     if (!isAnonymousUser()) {
-        signOut();
-        redirectToSignIn();
+        // We'll only call signOut if we're not stashing the session and this is not a supportal session,
+        // otherwise we'll call the API to invalidate the autogenerated credentials used for infinite
+        // session. SignOut also clears the storage, so we can just return after calling it.
+        const isSupportal = isSupportalToken();
+        if (!isSupportal && !stashSession) {
+            signOut();
+            return;
+        }
+
+        // The function redirectToSignIn will clear the whole storage, so let's create our onyx params
+        // updates for the credentials before we call it
+        let onyxSetParams = {};
+
+        // If we are not currently using a support token, and we received stashSession as true, we need to
+        // store the credentials so the user doesn't need to login again after they finish their supportal
+        // action. This needs to computed before we call `redirectToSignIn`
+        if (!isSupportal && stashSession) {
+            onyxSetParams = {
+                [ONYXKEYS.STASHED_CREDENTIALS]: credentials,
+                [ONYXKEYS.STASHED_SESSION]: session,
+            };
+        }
+        // If this is a supportal token, and we've received the parameters to stashSession as true, and
+        // we already have a stashedSession, that means we are supportaled, currently supportaling
+        // into another account and we want to keep the stashed data from the original account.
+        if (isSupportal && stashSession && hasStashedSession()) {
+            onyxSetParams = {
```

Comment:
They're necessary because `redirectToSignIn` will also clear up the whole storage. So we're waiting for it to clear everything, then we set those properties again. 

Having them cleared up is what guarantees that we won't have any "dirty" data from a different account.